src/net_processing.cpp

```diff
@@ -3879,39 +3992,35 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
         //
         // Message: getdata (non-blocks)
         //
+        // First check to see if there's anything in our notfound queue to request.
+        static uint64_t not_found_counter = 0;
+        ++not_found_counter;
+        if (!state.m_tx_download.m_not_found.empty()) {
+            // Check to see if it's our turn to drain the NOTFOUND queue.
+            // The idea is that we would like to randomly assign NOTFOUND
+            // transactions' next download request uniformly at random to our
+            // outbound peers, even though we don't necessarily know which ones
+            // will have the transaction available to download and not be up
+            // against their in-flight limit (or any other future restrictions
+            // we might impose).
+            // Our algorithm is to keep a global SendMessages() counter and
+            // every outbound peer gets to drain its notfound queue when the
+            // counter matches its location.
+            assert(g_outbound_peers.size() > 0);
+            if (g_outbound_peers[not_found_counter % g_outbound_peers.size()] == pto->GetId()) {
```

Comment:
Correct me if I'm wrong here, but isn't this equivalent to doing something like
```cpp
if (GetRandInt(/* nMax= */g_outbound_peers.size()) == 0) {
...
}
```
?

If you think the potential performance difference (which IMO is probably negligible when shoulder-to-shoulder with network IO) is okay, just doing something like that would allow you to keep `g_outbound_peers` a set and preserve the algorithm.

Reply:
Doing a 1 in N draw each time would introduce a bias to peers that are earlier in the loop, versus ones that are later in the loop.  Imagine if N=2, and we have two peers, peer0 and peer1.  peer0 would be selected with probability 1/2 + 1/8 + 1/16 + ... = 2/3.  If N=8, I think the first peer would have a 19% chance of being selected.

Reply:
I just realized that my hacky implementation here is broken too -- the `not_found_counter` will basically increment by the number of peers we have, so if the number of peers we have is equal the number of outbound peers we have (a common state for non-listening nodes) then we'll always be requesting notfound's from the first peer.

Reply:
I'm not sure I see how there's any bias based on the ordering of peers - the `GetRandInt` thing is an independent trial across peers, though of course we may end up draining multiple peers' `m_not_found` per single [`ThreadMessageHandler` loop](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L1954-L1978) if two peers both get `1/num_outbounds` lucky. Is there a reason we'd want to avoid that?

It seems like there might be a bias based upon [which peer we select to queue the `m_not_found` message on](https://github.com/bitcoin/bitcoin/pull/15505/files#diff-eff7adeaec73a769788bb78858815c91R3184) (peers earlier in `g_outbound_peers` will get selected first more often), but I'm not sure how the order in which we *flush* `m_not_found` makes a difference.

Reply:
Ok I redid this to be (hopefully) better, thanks.